crates/core/component/dex/src/component/candlestick/chandelier.rs

```diff
@@ -0,0 +1,429 @@
+    ) -> Result<()> {
+        let mut block_executions = self.block_executions_by_pair(trading_pair).clone();
+
+        // The execution occurred at the price of the previous state.
+        let execution_price = prev_state
+            .phi
+            .orient_start(trading_pair.end)
+            .context("position has one end = asset 2")?
+            .effective_price();
+
+        // The volume can be found by the change in reserves of the input asset.
+        let direct_volume = (new_state
+            .reserves_for(trading_pair.start)
+            .context("missing reserves")?
+            - prev_state
+                .reserves_for(trading_pair.start)
+                .context("missing reserves")?)
+        .into();
+
+        tracing::debug!(
+            ?trading_pair,
+            ?execution_price,
+            ?direct_volume,
+            "record position execution"
+        );
+        block_executions.push_back((execution_price, Some(direct_volume), None));
+        self.put_block_executions_by_pair(trading_pair, block_executions);
+
+        Ok(())
+    }
+
+    #[tracing::instrument(level = "debug", skip(self))]
+    async fn record_swap_execution(&mut self, swap: &SwapExecution) -> Result<()> {
+        let trading_pair: DirectedTradingPair = DirectedTradingPair {
+            start: swap.input.asset_id,
+            end: swap.output.asset_id,
+        };
+        let mut block_executions = self.block_executions_by_pair(&trading_pair).clone();
+
+        let execution_price = U128x128::ratio(swap.output.amount, swap.input.amount)
+            .context("denom unit is not 0")?;
+
+        // The volume is the amount of the input asset.
+        let swap_volume = swap.input.amount.into();
+
+        tracing::debug!(
+            ?trading_pair,
+            ?execution_price,
+            ?swap_volume,
+            "record swap execution"
+        );
+        block_executions.push_back((execution_price, None, Some(swap_volume)));
+        self.put_block_executions_by_pair(&trading_pair, block_executions);
+
+        Ok(())
+    }
+
+    #[tracing::instrument(level = "debug", skip(self))]
+    async fn finalize_block_candlesticks(&mut self) -> Result<()> {
+        let height = self.get_block_height().await?;
+
+        // Fetch all the executions for the block.
+        let block_executions = self.block_executions();
+
+        for (trading_pair, block_executions) in block_executions.iter() {
+            // Since the block executions are stored in order as they occurred during the block,
+            // we can iterate through them to create the candlestick data.
+            let mut open = None;
+            let mut close = 0.0;
+            let mut low = f64::INFINITY;
+            let mut high = 0.0;
+            let mut swap_volume = 0.0;
+            let mut direct_volume = 0.0;
+
+            // Create summary data on a per-trading pair basis.
+            for execution in block_executions {
```

Comment:
Q: do we know that every `block_executions` that appears here will be nonempty? In other words, do we have to worry about whether our sentinel values like `f64::INFINITY` would leak through into the data we produce?

My assumption is that we wouldn't end up with an empty `block_executions` but wanted to check.

Reply:
I don't believe it should be possible but I can add some checks to the `record_position_execution` and `record_swap_execution` to make sure that none are included.

Reply:
https://github.com/penumbra-zone/penumbra/blob/d6299e9dd1519784953139a6a2c37512239fdfe4/crates/core/component/dex/src/component/position_manager.rs#L375-L376  -- should be impossible for empty position executions to be recorded

Reply:
It does look like it's possible for no-op swap executions to be recorded, so an additional check there would be useful.